lib/grpc/stream.ex

```diff
@@ -1,17 +1,433 @@
+  """
+  @spec run_with(t(), Stream.t(), Keyword.t()) :: :ok | any()
+  def run_with(
+        %__MODULE__{flow: flow, options: flow_opts} = _stream,
+        %Materializer{} = from,
+        opts \\ []
+      ) do
+    dry_run? = Keyword.get(opts, :dry_run, false)
+    unary? = Keyword.get(flow_opts, :unary, false)
+
+    if unary? do
+      flow
+      |> Enum.to_list()
+      |> List.flatten()
+      |> List.first()
+    else
+      flow
+      |> Flow.map(fn msg ->
+        unless dry_run?, do: send_response(from, msg)
+        flow
+      end)
+      |> Flow.run()
+    end
+  end
+
+  @doc """
+  Sends a request to an external process and awaits a response.
+
+  If `target` is a PID, a message in the format `{:request, item, from}` is sent, and a reply
+  in the format `{:response, msg}` is expected.
+
+  If `target` is a GenServer module (atom), the message is `{:request, item}`, and a reply
+  in the format `{:response, msg}` is expected.
+
+  ## Parameters
+
+    - `stream`: The `GRPC.Stream` pipeline.
+    - `target`: Target process PID or GenServer module name.
+    - `timeout`: Timeout in milliseconds (defaults to `5000`).
+
+  ## Returns
+
+    - Updated stream if successful.
+    - `{:error, item, reason}` if the request fails or times out.
+  """
+  @spec ask(t(), pid | atom, non_neg_integer) :: t() | {:error, item(), reason()}
+  defdelegate ask(stream, target, timeout \\ 5000), to: Operators
+
+  @doc """
+  Same as `ask/3`, but raises an exception on failure.
+
+  ## Caution
+
+  This version propagates errors via raised exceptions, which can crash the Flow worker and halt the pipeline.
+  Prefer `ask/3` for production use unless failure should abort the stream.
+  """
+  @spec ask!(t(), pid | atom, non_neg_integer) :: t()
+  defdelegate ask!(stream, target, timeout \\ 5000), to: Operators
+
+  @doc """
+  Replaces the stream with a new one generated by the given factory function.
+  The factory function receives the metadata and the current item as arguments.
+  The factory function should return a new `GRPC.Stream` struct.
+
+  ## Caution
+  This function materializes the current stream and replaces it with a new one.
+  This can lead to performance issues if used excessively. In fact, it is more commonly used
+  in following the `single/2` of `single_as_ctx/3` functions. 
+
+  ## Parameters
+
+    - `stream`: The current `GRPC.Stream`.
+    - `factory`: A function that takes metadata and the current item, returning a new `GRPC.Stream`.
+
+  ## Returns
+    - A new `GRPC.Stream` struct with the replaced flow.
+
+  ## Example
+
+    GRPC.Stream.single(request)
+    # Create new output stream from the request and drop the input
+    |> GRPC.Stream.replace(fn _metadata, msg ->
+      create_output_stream(msg) # returns some Enumerable
+      |> GRPC.Stream.from()
+    end)
+    |> GRPC.Stream.run_with(materializer)
+  """
+  @spec replace(t(), (map(), term -> term)) :: t()
+  defdelegate replace(stream, factory), to: Operators
```

Comment:
I'm not sure I understand the usefulness of this function

Reply:
A server-side request means that after a client sends a single message a finite or infinite set of messages will be returned by the server to the client. This function allows this to be achieved in a more idiomatic way and with fewer steps for the library user. See the example in examples/helloworld_streams for more clarity.

Reply:
After reading the example, I think we should just remove this function. It is doing too much at the same time. The same results can be obtained by:

- Getting the result from the stream which you want to replace (your examples are both unary, which indicate you only want to deal with one result).
- Likewise, get the metadata and options from the same result
- Create a new Flow and GRPC.Stream as you would create one from scratch.

Reply:
I renamed the function to `via` and improved the doc

Reply:
I'm going to kindly ask you to keep this one. I really want to avoid a lot of unnecessary steps and make life easier for the user. Can you keep it?

Reply:
> After reading the example, I think we should just remove this function. It is doing too much at the same time. The same results can be obtained by:
> 
> * Getting the result from the stream which you want to replace (your examples are both unary, which indicate you only want to deal with one result).
> * Likewise, get the metadata and options from the same result
> * Create a new Flow and GRPC.Stream as you would create one from scratch.

Ok. I read it again and removed the function. I also updated the example.